### PR TITLE
Make sure to release connection lock

### PR DESCRIPTION
**Pull request by mikaelm12**
This is specifically to fix https://github.com/aspnet/AspNetCore-Internal/issues/2465
When acquiring the lock, if the connection wasn't in an active state we would get the lock and not release it. This was causing the timeouts when we tried to dispose.

### CODE REVIEW

File: src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs [hunk 1]
```diff
@@ -1948,6 +1948,7 @@ public async Task<ConnectionState> WaitForActiveConnectionAsync(string methodNam
 
                 if (CurrentConnectionStateUnsynchronized == null || CurrentConnectionStateUnsynchronized.Stopping)
                 {
+                    ReleaseConnectionLock();
```
Comment (BrennanConroy) [inline src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs]: Add `methodName`

Comment (BrennanConroy) [reply src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs]: Add a comment to the method about not putting it in a try/finally

Comment (BrennanConroy) [reply src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs]: Remove the flaky test attribute!

**Review by BrennanConroy — APPROVED** [review]

**Review by analogrelay — APPROVED** [review]

File: src/SignalR/clients/csharp/Client/test/FunctionalTests/HubConnectionTests.cs [hunk 2]
```diff
@@ -770,7 +770,6 @@ async IAsyncEnumerable<string> clientStreamData()
         }
 
         [Theory]
-        [Flaky("https://github.com/aspnet/AspNetCore-Internal/issues/2465", FlakyOn.All)]
         [MemberData(nameof(HubProtocolsAndTransportsAndHubPaths))]
```
Comment (jkotalik) [inline src/SignalR/clients/csharp/Client/test/FunctionalTests/HubConnectionTests.cs]: nit: consider adding [Repeat(int)] here to help diagnose faster if the test is flaky.

Comment (mikaelm12) [reply src/SignalR/clients/csharp/Client/test/FunctionalTests/HubConnectionTests.cs]: Talked irl

**Pull request merged**